### ISSUE TITLE
Support for syntax extensions

### ISSUE THREAD

**Issue by FynnFreyer** [DCsunset/pandoc-include#56]
# Transclusion Syntax

There's often lively discussion on how to do transclusion syntax in markdown. This old [discussion](https://talk.commonmark.org/t/transclusion-or-including-sub-documents-for-reuse/270) on the commonmark forum gives a decent overview.

A few popular ideas coming are:

1. `path/to/file.md` Just pasting the URL (from [iA Writer](https://github.com/iainc/Markdown-Content-Blocks))
2. `![description](path/to/file.md)` Repurposing image syntax (from [Obsidian](https://help.obsidian.md/Linking+notes+and+files/Embed+files))
3. `{{path/to/file.md}}` Using double braces (from [MultiMarkdown](https://fletcher.github.io/MultiMarkdown-4/transclusion))
4. `{{#include path/to/file.md}}` Double braces plus keyword (from [mdBook](https://rust-lang.github.io/mdBook/format/mdbook.html#including-files), also see #34)

Supporting a plethora of different syntaxes in this repo is probably not maintainable, and would complicate development. An alternative would be to specify an API for defining include syntaxes and then to allow users to extend syntax support via that API.

## Status Quo

Right now `pandoc-include` only supports the ``!include`<args>` path/to/file`` syntax, which has the advantage, that it allows for passing arguments to the directive, like you would in reST and that it's unambiguous. But it is a bit verbose and doesn't offer graceful degradation in other contexts.

The maintainer is open to a PR on this (see discussion in #54)

**Comment by FynnFreyer** [comment]
# Block or Inline Includes?

Should this inclusion syntax only work for including "blocks" of content? "Block" meaning a paragraph containing only a single element that conforms to some criterion as specified by the syntax implementation. Or should it also work for inline elements, like the image syntax?

I feel like only supporting blocks is reasonable, because the alternative would necessitate inlining the file contents, which seems kinda complicated for little to no benefit.

**Comment by FynnFreyer** [comment]
# Background

It seems like one could implement support for multiple syntaxes by adding an `is_include_line` function^[1] for each new syntax to support.

Right now function effectively has a signature of `is_include_line(elem: Para) -> tuple[int, str, dict]`, where the `int` is the include type (may as well be an `Enum`), the `str` is the file name, and the `dict` contains config values.^[2]

[1]: Maybe it also makes sense to allow extending `is_code_include`, but that calls `is_include_line` internally, so maybe it's unnecessary.
[2]: `is_code_include` has same signature and semantics as `is_include_line`

**Comment by FynnFreyer** [comment]
# Extension mechanism

A few popular syntaxes could be supported by default and the rest specified by users, but how would `pandoc-inline` know about a user defined syntax?

## External Packages

Users could implement new syntaxes in external packages.

We could then use [dynamic imports](https://docs.python.org/3/library/importlib.html#importlib.import_module) to load third party `is_include_line` functions.

Extensions could be [namespace packages](https://packaging.python.org/en/latest/guides/packaging-namespace-packages/), that reside under a common prefix like `pandoc-include-ext.pkgname`.

This is quite common in Sphinx extensions, c.f. the [`sphinxcontrib`](https://pypi.org/search/?q=sphinxcontrib&o=&c=Framework+%3A%3A+Sphinx+%3A%3A+Extension) namespace.

A meta-data option could then be used to activate an installed syntax.

## Pass Syntax Rules as Options

Another approach would be to just pass all the necessary information via the metadata options.

It seems that the heart of the include-recognition logic is in `is_include_line` and `extract_info`, which heavily rely on the regexes `RE_IS_INCLUDE_LINE`, `RE_IS_INCLUDE_HEADER`, and `RE_INCLUDE_PATTERN`. Maybe it's possible to define a new syntax by just passing these three regexes.

**Comment by FynnFreyer** [comment]
Both approaches have advantages and disadvantages.

External packages give much more control over the logic, but also require much more effort to implement. Passing regexes via metadata seems much simpler both for developing and using it, but has limited extensibility, and seems like kind of an ugly solution (hard to debug and full of confusing regex black magic incantations)

**Comment by FynnFreyer** [comment]
@DCsunset Are you okay with using the external package mechanism for extensions and limiting usage to block includes?

If so, I'd start looking into this

**Comment by DCsunset** [comment]
Yes I'm okay with it. Thanks for your interest in looking into it!

**Comment by FynnFreyer** [comment]
@DCsunset Last point is important and requires feedback

**Comment by DCsunset** [comment]
@FynnFreyer 
I think your changes look good to me generally. As for the last point, I think one solution is to pass all types of elememnts to the syntax module so the module can decide how to handle it. (it can simply return invalid if it's not an include element)

It's okay to change stuff in the action function or other places as you see fit. This will be included in a major release so it's a good opportunity to refactor some old stuff (while trying to minimize the changes perceivable by end users).

**Comment by DCsunset** [comment]
For the namespace package part, I think maybe we can allow using external packages by making the second parameters of `import_module` customizable as well in the future. In this way, the syntax package doesn't have to reside in this repo.
